cmd/kube-scheduler/app/server.go

```diff
@@ -194,25 +194,33 @@ func Run(ctx context.Context, cc *schedulerserverconfig.CompletedConfig, sched *
 		}
 	}
 
-	// Start all informers.
-	cc.InformerFactory.Start(ctx.Done())
-	// DynInformerFactory can be nil in tests.
-	if cc.DynInformerFactory != nil {
-		cc.DynInformerFactory.Start(ctx.Done())
-	}
+	startInformersAndWaitForSync := func() {
+		// Start all informers.
+		cc.InformerFactory.Start(ctx.Done())
+		// DynInformerFactory can be nil in tests.
+		if cc.DynInformerFactory != nil {
+			cc.DynInformerFactory.Start(ctx.Done())
+		}
 
-	// Wait for all caches to sync before scheduling.
-	cc.InformerFactory.WaitForCacheSync(ctx.Done())
-	// DynInformerFactory can be nil in tests.
-	if cc.DynInformerFactory != nil {
-		cc.DynInformerFactory.WaitForCacheSync(ctx.Done())
+		// Wait for all caches to sync before scheduling.
+		cc.InformerFactory.WaitForCacheSync(ctx.Done())
+		// DynInformerFactory can be nil in tests.
+		if cc.DynInformerFactory != nil {
+			cc.DynInformerFactory.WaitForCacheSync(ctx.Done())
+		}
+	}
+	if cc.ComponentConfig.MaintainCacheWhenNonLeader {
+		startInformersAndWaitForSync()
 	}
 
 	// If leader election is enabled, runCommand via LeaderElector until done and exit.
 	if cc.LeaderElection != nil {
 		cc.LeaderElection.Callbacks = leaderelection.LeaderCallbacks{
 			OnStartedLeading: func(ctx context.Context) {
 				close(waitingForLeader)
+				if !cc.ComponentConfig.MaintainCacheWhenNonLeader {
+					startInformersAndWaitForSync()
```

Comment:
what guards against a scheduler obtaining and keeping the leader lease, failing to start/sync informers here, and hanging indefinitely while holding the leader lease?

Reply:
AIUI, this is not a problem brought in by this change, right?

Reply:
It is, because previously a scheduler only requests a lease after the informers already synced.

Reply:
Right, I checked kube-controller-manager code and it should have same issue since it starts informers after having got the lease. I can't think of a nice solution to it but how likely are informers to fail to start/sync? would that indicate a bigger problem that scheduler can't handle anyway?

Reply:
In theory, this will also happen with scheduler, right?

Reply:
Do you think we should have a mechanism to make leader election aware of the failure in `OnStartedLeading` and then release the lease.

Reply:
Can you please elaborate on the failures we have here?

I had a quick read over `startInformersAndWaitForSync` and `sched.run()`. Both of them don't return any explicit errors but just return to complete the function. I can think of one possible failure in the informers where they stuck to pull resource objects, but I don't see a mechanism in informer APIs to capture that.

Alternatively, we can have a new `context` with timeout passed to informers. However, it brings up a new question as to how long the timeout should be.